Fix API importing scripts

Drulex opened the pull request:
Closes #2963 (almost... there still remains `spinalcordtoolbox.moco` that needs some more love - will handle in a separate issue).

- Remove script imports from API modules
- Refactor `flatten_sagittal` into `spinalcordtoolbox.flattening` since it is needed elsewhere in the API (`detect_c2c3`)
- Move `scripts` inside package so that code that legitimately needs to import scripts (e.g unit tests, but not API code) can import cleanly without modifying the `PATH ` variable.

Batch processing results (2020-11-30):
~~~
Version:         git-aj/2963-fix-api-importing-scripts-bb32a7b5fbbd65112ea827f56e1bcd4d245ca527
Ran on:          Linux gentoo-p53 5.4.80-gentoo-r1-x86_64
Duration:        0hrs 11min 21sec
---
t2/CSA:          73.87711295363036
mt/MTR(WM):      54.38328519535009
t2s/CSA_GM:      12.487834828856178
t2s/CSA_WM:      64.93830702088246
dmri/FA(CST_r):  0.7756265311035226
dmri/FA(CST_l):  0.7800112266489608
~~~

Drulex commented:
Functional tests live outside the package, but will need to import from `scripts` so I think that it's reasonable for them to use the `sys.path.append(os.path.join(__sct_dir__, 'scripts'))` trick, unless there is a better/cleaner way to do this? Thoughts @joshuacwnewton, @kousu ?

jcohenadad commented:
> Functional tests live outside the package, but will need to import from scripts so I think that it's reasonable for them to use the sys.path.append(os.path.join(__sct_dir__, 'scripts')) trick, unless there is a better/cleaner way to do this? Ideas @joshuacwnewton, @kousu ?

is that a crazy idea to include `scripts/` inside the package? I've seen a couple of Python projects where "scripts/" or "cli/" are included.

Drulex commented:
> is that a crazy idea to include `scripts/` inside the package? I've seen a couple of Python projects where "scripts/" or "cli/" are included.

Honestly I don't know what the best practice is here. Instinctively I'm thinking `scripts/cli` are not part of the library so it makes sense that they would be outside the package. However it makes it no-obvious how to import them in tests, which is a reasonable use case. Also from the zen of python:
> There should be one-- and preferably only one --obvious way to do it.
> Although that way may not be obvious at first unless you're Dutch.

I don't see a downside of putting them in the package, as long as import hierarchy is still respected.

This article (https://docs.python-guide.org/writing/structure/#test-suite), proposes to have a `tests/context.py` that imports the package for use cases where someone would want to run tests without having the package installed. I guess this logic could be re-used in for scripts in the following way:

```python
# unit_testing/cli/context.py

import os
import sys
sys.path.append(os.path.join(${sct_dir}, 'scripts'))

# import all scripts
import sct_something
# [...]
```
And then the functional tests can be written like this:

```python

# unit_testing/cli/test_cli_sct_something.py
from .context import sct_something
```

joshuacwnewton commented on unit_testing/scripts.py:
```diff
@@ -0,0 +1,63 @@
+# add scripts location to path
+sys.path.append(os.path.join(__sct_dir__, 'scripts'))
+
+# import all scripts
+import sct_analyze_lesion
```
I'm a little confused about why we're doing the imports here in this file. If we're adding `scripts/` to the path, can't we just do `from scripts import <sct_function>` directly in each test?

Drulex replied:
To avoid adding it to the path in each test file. Tests can be run individually therefore they cannot assume that `scripts` is in the path since it is not part of the package.

We can't do `from scripts import <sct_function>`, because we are adding the `scripts` directory itself to the path. Only the contents (i.e the scripts themselves) is accessible.

What I like about the proposed approach here is that it is explicit and straightforward. `scripts.py` exposes the scripts and the tests that need to import a script may just use it via relative import of `.scripts`.

joshuacwnewton replied:
> To avoid adding it to the path in each test file. Tests can be run individually therefore they cannot assume that scripts is in the path since it is not part of the package.

Would `conftest.py` be an appropriate place to put this to cover all of the tests? Actually... it's already there. :sweat_smile: 

https://github.com/neuropoly/spinalcordtoolbox/blob/c8eaa4a202deaaba72f080c7506e5bb0b144efa6/conftest.py#L23-L25

> We can't do from scripts import <sct_function>, because we are adding the scripts directory itself to the path. Only the contents (i.e the scripts themselves) is accessible.

Huh... this is a bit confusing to me. I tried `from scripts import <sct_function>` for a test and it worked? Perhaps I am misunderstanding something about Python imports.

Scripts currently has an `__init__.py` thanks to #2837 so I believe it's treated as its own package... perhaps that's why it's working for me?

> What I like about the proposed approach here is that it is explicit and straightforward. scripts.py exposes the scripts and the tests that need to import a script may just use it via relative import of .scripts.

My only concern is that it requires manual updating when a script is added/removed/etc. But apart from that, I agree.

Drulex replied:
> Would `conftest.py` be an appropriate place to put this to cover all of the tests? Actually... funny enough, it's already there. So, currently the scripts folder is added to the path twice.

It's there, but it shouldn't be. It's only there because `sct_download` is used in a pre-test collection hook and we haven't yet decided what the final form of `sct_download` should be.

> Huh... this is a bit confusing to me. I tried from scripts import <sct_function> for a test and it worked? Perhaps I am misunderstanding something about Python imports.

Probably because `$SCT_DIR` was added to `$PATH` somewhere else coincidentally. `scripts` doesn't have a parent package because the top-level `spinalcordtoolbox` folder doesn't have an `__init__.py`, otherwise we would have to do `from spinalcordtoolbox.scripts import sct_something` and `from spinalcordtoolbox.spinalcordtoolbox import some_api_module`.. If I launch a python shell in my venv and print `sys.path` I can see the `$SCT_DIR` in there whereas my regular path doesn't have it. In any case it is not the desired behavior.

> Scripts currently has an __init__.py thanks to #2837 so I believe it's treated as its own package... perhaps that's why it's working for me?

That's good, it allows a script to import another relative script via `from . import sct_something`. Same logic as what I am introducing in the `unit_testing` folder in this PR.

> My only concern is that it requires manual updating when a script is added/removed/etc

Yes and same goes for any API module once we merge https://github.com/neuropoly/spinalcordtoolbox/pull/2955/commits/f65df321b078b939e5de66755ca0561fe773ed82 because it allows more sane imports like `import spinalcordtoolbox as sct` with access to all sub-packages without explicit importing. For example you expect to do:
```
import os
os.path.join(...)
```
instead of:
```
import os
import os.path
import os.path.join
os.path.join(...)
```
Right now on `master` we can only do the second form, this PR combined with #2955 allows the first form which is standard practice.

joshuacwnewton replied:
Thanks for the clarification. :slightly_smiling_face:

Drulex replied:
The `os` example I gave is probably a bad one since `os` is part of the standard lib with an implementation scattered across multiple files. The `import numpy as np` and usage of `np` in general is a better one (https://github.com/numpy/numpy/blob/master/numpy/__init__.py)

Drulex commented:
I've run into a limitation with the method proposed: removing the `sys.path` hack and changing the scripts to use relative imports does not work because `scripts` don't have a parent package: `ImportError: attempted relative import with no known parent package` (which is obvious, I completely missed it before actually trying it).

> is that a crazy idea to include scripts/ inside the package? I've seen a couple of Python projects where "scripts/" or "cli/" are included.

Looks like it makes sense to include `scripts` in the package after all. Will do the change.

kousu commented:
I always find python3's package system (the installed system not the packag_ing_ system) so difficult. I think they designed it mostly with absolute imports in mind, and it has some weird quirks and limitations because they wanted to support loading packages from zip files or over the network too. The path you launch a script from will affect the pythonpath too, which means switching in or out of venvs can complicate things too.

joshuacwnewton reviewed (commented):
Looks straightforward to me! Only a few comments.

joshuacwnewton commented:
Oh, one other question (because the GitHub UI wouldn't let me highlight changes on empty files): 

Are `unit_testing/__init__.py` and `unit_testing/cli/__init__.py` needed?

Drulex commented on spinalcordtoolbox/__init__.py:
```diff
@@ -1,4 +1,21 @@
+# explicitely expose __version__ and friends in top-level package
 from .utils import __version__, __sct_dir__, __data_dir__, __deepseg_dir__
+
+from . import aggregate_slicewise
+from . import cropping
+from . import download
+from . import image
+from . import labels
+from . import math
+from . import metadata
+from . import flattening
+from . import process_seg
+from . import resampling
+from . import straightening
+from . import template
+from . import types
```
Importing a package implicitly runs the `__init__.py` inside the package. If you take a look at the sub-packages they also import their respective relative modules inside their `__init__.py`.

More info here: https://docs.python.org/3/reference/import.html#regular-packages

joshuacwnewton replied:
If I'm understanding correctly, though, when we run `import spinalcordtoolbox` then currently, `sct.deepseg`, etc. won't be available by default. Only what we've explicitly named here? 

```python
>>> import spinalcordtoolbox as sct
>>> sct.download
Out[3]: <module 'spinalcordtoolbox.download' from '/home/joshua/Repos/spinalcordtoolbox/spinalcordtoolbox/download.py'>
>>> sct.download.download_data
Out[4]: <function spinalcordtoolbox.download.download_data(urls)>
>>> sct.deepseg
Traceback (most recent call last):
  File "/home/joshua/Repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3343, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-deddbddf3fe5>", line 1, in <module>
    sct.deepseg
AttributeError: module 'spinalcordtoolbox' has no attribute 'deepseg'
```

This is what looks to be available with `import spinalcordtoolbox`:

![Screenshot from 2020-10-30 09-32-42](https://user-images.githubusercontent.com/16181459/97711003-e2f22700-1a92-11eb-8a12-06b34158220c.png)

Drulex replied:
```
(venv_sct) ➜  spinalcordtoolbox git:(aj/2963-fix-api-importing-scripts) git checkout master                           
Switched to branch 'master'
Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
(venv_sct) ➜  spinalcordtoolbox git:(master) python
Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import spinalcordtoolbox as sct
>>> dir(sct)
['__builtins__', '__cached__', '__data_dir__', '__deepseg_dir__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__sct_dir__', '__spec__', '__version__', 'utils']
>>> 
(venv_sct) ➜  spinalcordtoolbox git:(master) git checkout aj/2963-fix-api-importing-scripts
Switched to branch 'aj/2963-fix-api-importing-scripts'
(venv_sct) ➜  spinalcordtoolbox git:(aj/2963-fix-api-importing-scripts) python
Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import spinalcordtoolbox as sct
>>> dir(sct)
['__builtins__', '__cached__', '__data_dir__', '__deepseg_dir__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__sct_dir__', '__spec__', '__version__', 'aggregate_slicewise', 'centerline', 'cropping', 'download', 'flattening', 'image', 'labels', 'math', 'metadata', 'process_seg', 'resampling', 'straightening', 'template', 'types', 'utils']
```

Drulex replied:
I don't see how you have access to `sct.download` on `master` without a fully qualified import. Maybe because your environment isn't 100% clean? Maybe you have `spinalcordtoolbox/spinalcordtoolbox/` in your `PATH`? 

You can try in a fresh environment:
```
# from inside sct dir
docker run -v /tmp/spinalcordtoolbox:/spinalcordtoolbox --interactive --tty --entrypoint /bin/bash --net=host python:3.6
# run install_sct and try again?
```

joshuacwnewton replied:
I'm not on `master`? What you're getting [here](https://github.com/neuropoly/spinalcordtoolbox/pull/2966#discussion_r515104647) for both branches is what I get too. :)

That's sort of a different thing than what I was asking, heh.

Drulex replied:
We export the subpackages via their own `__init__.py`. Python looks recursively for `__init__.py` files when importing a package. If it sees an `__init__.py` in a sub-folder it runs it.

e.g: https://github.com/neuropoly/spinalcordtoolbox/blob/c8af6cc7bc6898ec33c9670a89cf04cafccdb6a6/spinalcordtoolbox/deepseg/__init__.py

joshuacwnewton replied:
But that means we can't import just `spinalcordtoolbox` and access `deepseg`, then? i.e. we have to explicitly import `deepseg` itself?

Drulex replied:
Before you couldn't, now you can (after this).

See relevant: https://github.com/neuropoly/spinalcordtoolbox/pull/2969#discussion_r511924214

^ same concept but a level higher

joshuacwnewton replied:
Sorry, I still don't think I have explained myself very well...

> See relevant: https://github.com/neuropoly/spinalcordtoolbox/pull/2969#discussion_r511924214

The usage there still requires you to import `deepseg` explicitly, using `import spinalcordtoolbox.deepseg`.

What I'm talking about is just importing `spinalcordtoolbox` by itself, then having access to the subpackages immediately. i.e. the following:

```python
>>> import spinalcordtoolbox as sct
>>> sct.deepseg
<module 'spinalcordtoolbox.deepseg' from '/home/joshua/Repos/spinalcordtoolbox/spinalcordtoolbox/deepseg/__init__.py'>
>>> sct.deepseg.models.list_models()
{'t2star_sc': {'url': ['https://github.com/ivadomed/t2star_sc/releases/download/r20200622/r20200622_t2star_sc.zip',
   'https://osf.io/v9hs8/download?version=5'],
  'description': 'Cord segmentation model on T2*-weighted contrast.',
  'default': True},
```

So, I am suggesting to possibly add `from . import deepseg` (along with the other subpackages) to the top-level package `__init__.py`.

zougloub replied:
The `import spinalcordtoolbox` having access to submodules immediately should be subject to discussion in another issue.

joshuacwnewton replied:
@zougloub Isn't giving access to submodules immediately the whole purpose of this PR? From @Drulex's https://github.com/neuropoly/spinalcordtoolbox/pull/2966#discussion_r511195929

> Yes and same goes for any API module once we merge [f65df32](https://github.com/neuropoly/spinalcordtoolbox/commit/f65df321b078b939e5de66755ca0561fe773ed82) because it allows more sane imports like `import spinalcordtoolbox as sct` with access to all sub-packages without explicit importing. For example you expect to do:
> 
> ```
> import os
> os.path.join(...)
> ```
> 
> instead of:
> 
> ```
> import os
> import os.path
> import os.path.join
> os.path.join(...)
> ```
> 
> Right now on `master` we can only do the second form, this PR combined with #2955 allows the first form which is standard practice.

Drulex replied:
Sorry @joshuacwnewton I misunderstood what you were saying. Indeed this PR does not propose to implicitly expose sub-packages, but only modules of the `spinalcordtoolbox` package. In the PR's current form sub-packages would still need to be imported explicitly. I created an issue to discuss since it seems to be a hot topic: #2989

Drulex commented on spinalcordtoolbox/deepseg_sc/core.py:
```diff
@@ -14,14 +14,11 @@
 from .cnn_models import nn_architecture_seg, nn_architecture_ctr
 from .postprocessing import post_processing_volume_wise, keep_largest_object, fill_holes_2d
-from spinalcordtoolbox.image import Image, empty_like, change_type, zeros_like, add_suffix
+from spinalcordtoolbox.image import Image, empty_like, change_type, zeros_like, add_suffix, concat_data, split_img_data
 from spinalcordtoolbox.centerline.core import ParamCenterline, get_centerline, _call_viewer_centerline
 from spinalcordtoolbox.utils import sct_dir_local_path, TempFolder
 from spinalcordtoolbox.deepseg_sc.cnn_models_3d import load_trained_model
 
-# FIXME: don't import from scripts!
-from sct_image import concat_data, split_data
```
`concat_data()` and `split_data()` were refactored to `image.py` at some point in the past, but not all callers were updated (simply due to scope of PR).

If you look at `sct_image.py` it maintains backwards compatibility for the other callers yet to be refactored, but internally calls the API.

https://github.com/neuropoly/spinalcordtoolbox/blob/c8af6cc7bc6898ec33c9670a89cf04cafccdb6a6/spinalcordtoolbox/scripts/sct_image.py#L188-L192

https://github.com/neuropoly/spinalcordtoolbox/blob/c8af6cc7bc6898ec33c9670a89cf04cafccdb6a6/spinalcordtoolbox/scripts/sct_image.py#L408-L412

Callers of `sct_image.py -concat` and `sct_image.py -split` can safely be refactored to use the `image` API. I should get there soon.

Drulex commented on spinalcordtoolbox/scripts/sct_create_mask.py:
```diff
@@ -30,7 +30,7 @@
 from spinalcordtoolbox.types import Coordinate
 
-from sct_image import concat_data
+from spinalcordtoolbox.scripts.sct_image import concat_data
```
Created https://github.com/neuropoly/spinalcordtoolbox/projects/7#card-48403943

joshuacwnewton replied:
Thanks! :)

Drulex commented:
> Oh, one other question (because the GitHub UI wouldn't let me highlight changes on empty files):
> 
> Are `unit_testing/__init__.py` and `unit_testing/cli/__init__.py` needed?

Maybe `unit_testing/cli/__init__.py` isn't needed, but `unit_testing/__init__.py` is because some unit tests import fixtures from others. e.g:

https://github.com/neuropoly/spinalcordtoolbox/blob/c8af6cc7bc6898ec33c9670a89cf04cafccdb6a6/unit_testing/test_labels.py#L14

`unit_testing/` is not part of the top-level package so I it couldn't cause import issues when using `sct`.

joshuacwnewton commented:
>  but `unit_testing/__init__.py` is because some unit tests import fixtures from others. e.g:

Hmm, this makes me wonder about pytest's recommendations on this front. See the description for [`conftest.py`](https://docs.pytest.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions) and the section below it in the docs. But, that's out of scope of this PR, I think.

zougloub commented:
@Drulex IMHO the import spinalcordtoolbox having access to submodules immediately should be subject to discussion in another issue.

jcohenadad commented:
Do we need this PR for v5.0.0?

Drulex commented:
Reviving this PR.

Based on the discussions in this thread I removed all code the implicitly imports sub-packages to avoid the performance impacts.

I've updated the description to reflect the scope of the PR.

zougloub reviewed (approved):
LGTM

joshuacwnewton commented:
I think it would be good to run CI before merging, as soon as we resolve the Travis issues.
